File: apps/api/services/billing_customer_sync.py
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from core.id import generate_id
from db.models import BillingCustomerOutbox, Membership, Organization, User

CUSTOMER_EVENT_TYPE = "customer.ensure"


@dataclass(frozen=True)
class PartySnapshot:
    """The identity fields Billing needs to render a customer without a callback."""

    subject_type: str
    subject_id: str
    customer_kind: str
    name: str
    email: str | None
    company_name: str | None
    first_name: str | None
    last_name: str | None
    phone: str | None
    contact_user_id: str | None
    contact_first_name: str | None
    contact_last_name: str | None
    contact_email: str | None
    contact_phone: str | None


def _full_name(first_name: str | None, last_name: str | None) -> str | None:
    parts = [part.strip() for part in (first_name, last_name) if part and part.strip()]
    return " ".join(parts) or None


def _customer_name_for_user(user: User) -> str:
    return (
        getattr(user, "name", None)
        or _full_name(getattr(user, "first_name", None), getattr(user, "last_name", None))
        or getattr(user, "username", None)
        or getattr(user, "email", None)
        or user.id
    )


async def _resolve_org_primary_contact(
    db: AsyncSession,
    organization: Organization,
) -> User | None:
    """Resolves an organization's primary contact.

    Order: the organization's declared `primary_contact_user_id`, then the
    owner membership, then the earliest membership. Returns None when the org
    has no members — callers must not fabricate a contact.
    """
    declared_id = getattr(organization, "primary_contact_user_id", None)
    if declared_id:
        declared = await db.get(User, declared_id)
        if declared is not None:
            return declared

    memberships = list(
        (
            await db.scalars(
                select(Membership)
                .where(Membership.organization_id == organization.id)
                .order_by(Membership.created_at, Membership.id)
            )
        ).all()
    )
    if not memberships:
        return None

    owner = next((m for m in memberships if m.role == "owner"), memberships[0])
    return await db.get(User, owner.user_id)


async def snapshot_for_organization(db: AsyncSession, organization: Organization) -> PartySnapshot:
    """Builds the BUSINESS party snapshot for a core organization."""
    contact = await _resolve_org_primary_contact(db, organization)
    legal_name = organization.name or organization.id
    display_name = getattr(organization, "doing_business_as", None) or legal_name

    return PartySnapshot(
        subject_type="organization",
        subject_id=organization.id,
        customer_kind="BUSINESS",
        name=display_name,
        email=getattr(organization, "primary_email", None) or (contact.email if contact else None),
        company_name=legal_name,
        # An organization's first/last name mirror its primary contact so a
        # single "customer name" column renders a person for both party kinds.
        first_name=contact.first_name if contact else None,
        last_name=contact.last_name if contact else None,
        phone=getattr(organization, "primary_phone", None),
        contact_user_id=contact.id if contact else None,
        contact_first_name=contact.first_name if contact else None,
        contact_last_name=contact.last_name if contact else None,
        contact_email=contact.email if contact else None,
        contact_phone=getattr(contact, "phone", None) if contact else None,
    )


def snapshot_for_user(user: User) -> PartySnapshot:
    """Builds the INDIVIDUAL party snapshot for a core user."""
    return PartySnapshot(
        subject_type="user",
        subject_id=user.id,
        customer_kind="INDIVIDUAL",
        name=_customer_name_for_user(user),
        email=getattr(user, "email", None) or None,
        company_name=None,
        first_name=getattr(user, "first_name", None) or None,
        last_name=getattr(user, "last_name", None) or None,
        phone=getattr(user, "phone", None),
        # A person is their own contact; Billing seeds no separate contact row.
        contact_user_id=None,
        contact_first_name=None,
        contact_last_name=None,
        contact_email=None,
        contact_phone=None,
    )


def _payload_hash(snapshot: PartySnapshot) -> str:
    payload = json.dumps(snapshot.__dict__, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


async def _enqueue_customer_ensure(
    db: AsyncSession,
    snapshot: PartySnapshot,
    now: int,
) -> bool:
    payload_hash = _payload_hash(snapshot)

    latest = (
        await db.scalars(
            select(BillingCustomerOutbox)
            .where(
                BillingCustomerOutbox.subject_type == snapshot.subject_type,
                BillingCustomerOutbox.subject_id == snapshot.subject_id,
            )
            .order_by(BillingCustomerOutbox.created_at.desc(), BillingCustomerOutbox.id.desc())
            .limit(1)
        )
    ).first()

    if latest is not None:
        # An undelivered event is refreshed in place rather than duplicated.
        if latest.status in ("pending", "processing"):
            if latest.payload_hash != payload_hash:
                _apply_snapshot(latest, snapshot, payload_hash, now)
                latest.updated_at = now
                await db.flush()
            return False
        # Nothing changed since the last successful delivery.
        if latest.status == "delivered" and latest.payload_hash == payload_hash:
            return False

    event = BillingCustomerOutbox(
        id=generate_id("billingCustomerEvent"),
        event_type=CUSTOMER_EVENT_TYPE,
        subject_type=snapshot.subject_type,
        subject_id=snapshot.subject_id,
        name=snapshot.name,
        email=snapshot.email,
        occurred_at=now,
        status="pending",
        attempt_count=0,
        available_at=now,
        locked_at=None,
        delivered_at=None,
        last_error=None,
        created_at=now,
        updated_at=now,
    )
    _apply_snapshot(event, snapshot, payload_hash, now)
    db.add(event)
    await db.flush()
    return True


def _apply_snapshot(
    event: BillingCustomerOutbox,
    snapshot: PartySnapshot,
    payload_hash: str,
    now: int,
) -> None:
    event.name = snapshot.name
    event.email = snapshot.email
    event.customer_kind = snapshot.customer_kind
    event.company_name = snapshot.company_name
    event.first_name = snapshot.first_name
    event.last_name = snapshot.last_name
    event.phone = snapshot.phone
    event.contact_user_id = snapshot.contact_user_id
    event.contact_first_name = snapshot.contact_first_name
    event.contact_last_name = snapshot.contact_last_name
    event.contact_email = snapshot.contact_email
    event.contact_phone = snapshot.contact_phone
    event.payload_hash = payload_hash
    event.occurred_at = now


async def enqueue_customer_ensure_for_organization(
    db: AsyncSession,
    organization: Organization,
    now: int,
) -> None:
    await _enqueue_customer_ensure(db, await snapshot_for_organization(db, organization), now)


async def enqueue_customer_ensure_for_user(db: AsyncSession, user: User, now: int) -> None:
    await _enqueue_customer_ensure(db, snapshot_for_user(user), now)


def customer_event_payload(event: BillingCustomerOutbox) -> dict[str, object]:
    """Serializes one outbox row into the Billing `customer.ensure` contract."""
    common: dict[str, object] = {
        "customerKind": event.customer_kind
        or ("BUSINESS" if event.subject_type == "organization" else "INDIVIDUAL"),
        "name": event.name,
        "email": event.email,
        "firstName": event.first_name,
        "lastName": event.last_name,
        "phone": event.phone,
    }

    if event.subject_type == "organization":
        return {
            **common,
            "customerType": "CORE_ORGANIZATION",
            "organizationId": event.subject_id,
            "companyName": event.company_name,
            "primaryContact": _contact_payload(event),
        }

    return {
        **common,
        "customerType": "CORE_USER",
        "userId": event.subject_id,
    }


def _contact_payload(event: BillingCustomerOutbox) -> dict[str, object] | None:
    if not event.contact_user_id:
        return None
    return {
        "userId": event.contact_user_id,
        "firstName": event.contact_first_name,
        "lastName": event.contact_last_name,
        "email": event.contact_email,
        "phone": event.contact_phone,
    }


async def enqueue_reconcile_all(db: AsyncSession, now: int) -> dict[str, int]:
    """Re-enqueues every organization, plus every user already known to Billing.

    Organizations are always reconciled: an org is a customer of the platform
    from the moment it exists. Users are only reconciled when a prior event
    exists for them — signing up for a free 876 account does not make someone a
    customer, so reconciliation must not manufacture one.
    """
    organizations = list((await db.scalars(select(Organization).order_by(Organization.id))).all())
    organization_count = 0
    for organization in organizations:
        if await _enqueue_customer_ensure(db, await snapshot_for_organization(db, organization), now):
            organization_count += 1

    known_user_ids = list(
        (
            await db.scalars(
                select(BillingCustomerOutbox.subject_id)
                .where(BillingCustomerOutbox.subject_type == "user")
                .distinct()
            )
        ).all()
    )
    user_count = 0
    for user_id in known_user_ids:
        user = await db.get(User, user_id)
        if user is None:
            continue
        if await _enqueue_customer_ensure(db, snapshot_for_user(user), now):
            user_count += 1

    return {"organizations": organization_count, "users": user_count}

File: apps/api/src/services/billing-customer-sync.ts
import { createHash } from 'node:crypto'

import { generateId } from '@/platform/ids'

export const CUSTOMER_EVENT_TYPE = 'customer.ensure'

export type PartySnapshot = {
  subjectType: string
  subjectId: string
  customerKind: string
  name: string
  email: string | null
  companyName: string | null
  firstName: string | null
  lastName: string | null
  phone: string | null
  contactUserId: string | null
  contactFirstName: string | null
  contactLastName: string | null
  contactEmail: string | null
  contactPhone: string | null
}

export type OrganizationRow = {
  id: string
  name: string | null
  shortName?: string | null
  slug: string
  doingBusinessAs?: string | null
  primaryEmail?: string | null
  primaryPhone?: string | null
  primaryContactUserId?: string | null
}

export type UserRow = {
  id: string
  email: string | null
  name?: string | null
  firstName?: string | null
  lastName?: string | null
  username?: string | null
  phone?: string | null
}

export type MembershipRow = {
  id: string
  organizationId: string
  userId: string
  role: string
  createdAt: bigint
}

export type BillingCustomerOutboxRow = {
  id: string
  eventType: string
  subjectType: string
  subjectId: string
  name: string
  email: string | null
  occurredAt: bigint
  status: string
  attemptCount: number
  availableAt: bigint
  lockedAt: bigint | null
  deliveredAt: bigint | null
  lastError: string | null
  createdAt: bigint
  updatedAt: bigint
  customerKind: string | null
  companyName: string | null
  firstName: string | null
  lastName: string | null
  phone: string | null
  contactUserId: string | null
  contactFirstName: string | null
  contactLastName: string | null
  contactEmail: string | null
  contactPhone: string | null
  payloadHash: string | null
}

export type BillingCustomerSyncRepository = {
  findUserById(userId: string): Promise<UserRow | null>
  listMembershipsByOrganizationId(
    organizationId: string
  ): Promise<MembershipRow[]>
  findLatestOutboxBySubject(
    subjectType: string,
    subjectId: string
  ): Promise<BillingCustomerOutboxRow | null>
  createOutboxEvent(
    data: BillingCustomerOutboxRow
  ): Promise<BillingCustomerOutboxRow>
  updateOutboxEvent(
    eventId: string,
    data: Partial<BillingCustomerOutboxRow>
  ): Promise<BillingCustomerOutboxRow>
  listOrganizations(): Promise<OrganizationRow[]>
  listKnownUserIds(): Promise<string[]>
}

export type BillingCustomerSyncDeps = {
  repository: BillingCustomerSyncRepository
}

function fullName(
  firstName: string | null | undefined,
  lastName: string | null | undefined
): string | null {
  const parts = [firstName, lastName]
    .filter(
      (part): part is string =>
        typeof part === 'string' && part.trim().length > 0
    )
    .map((part) => part.trim())
  return parts.length > 0 ? parts.join(' ') : null
}

function customerNameForUser(user: UserRow): string {
  return (
    (user.name?.trim() || null) ??
    fullName(user.firstName, user.lastName) ??
    (user.username?.trim() || null) ??
    (user.email?.trim() || null) ??
    user.id
  )
}

async function resolveOrgPrimaryContact(
  repository: BillingCustomerSyncRepository,
  organization: OrganizationRow
): Promise<UserRow | null> {
  const declaredId = (organization as Record<string, unknown>)
    .primaryContactUserId as string | null | undefined
  if (declaredId) {
    const declared = await repository.findUserById(declaredId)
    if (declared) return declared
  }

  const memberships = await repository.listMembershipsByOrganizationId(
    organization.id
  )
  if (memberships.length === 0) return null

  const sorted = [...memberships].sort((a, b) => {
    if (a.createdAt !== b.createdAt) return a.createdAt < b.createdAt ? -1 : 1
    return a.id.localeCompare(b.id)
  })

  const owner = sorted.find((m) => m.role === 'owner') ?? sorted[0]!
  return repository.findUserById(owner.userId)
}

export async function snapshotForOrganization(
  deps: BillingCustomerSyncDeps,
  organization: OrganizationRow
): Promise<PartySnapshot> {
  const contact = await resolveOrgPrimaryContact(deps.repository, organization)
  const legalName = organization.name ?? organization.id
  const displayName = (organization as Record<string, unknown>)
    .doingBusinessAs as string | null | undefined
  const name = (displayName as string | null) ?? legalName

  const primaryEmail = (organization as Record<string, unknown>)
    .primaryEmail as string | null | undefined
  const primaryPhone = (organization as Record<string, unknown>)
    .primaryPhone as string | null | undefined

  const contactPhone = contact
    ? (((contact as Record<string, unknown>).phone as
        | string
        | null
        | undefined) ?? null)
    : null

  return {
    subjectType: 'organization',
    subjectId: organization.id,
    customerKind: 'BUSINESS',
    name,
    email: (primaryEmail as string | null) ?? contact?.email ?? null,
    companyName: legalName,
    firstName: contact?.firstName ?? null,
    lastName: contact?.lastName ?? null,
    phone: (primaryPhone as string | null) ?? null,
    contactUserId: contact?.id ?? null,
    contactFirstName: contact?.firstName ?? null,
    contactLastName: contact?.lastName ?? null,
    contactEmail: contact?.email ?? null,
    contactPhone,
  }
}

export function snapshotForUser(user: UserRow): PartySnapshot {
  const email = (user.email?.trim() ? user.email : null) ?? null
  return {
    subjectType: 'user',
    subjectId: user.id,
    customerKind: 'INDIVIDUAL',
    name: customerNameForUser(user),
    email,
    companyName: null,
    firstName: (user.firstName?.trim() ? user.firstName : null) ?? null,
    lastName: (user.lastName?.trim() ? user.lastName : null) ?? null,
    phone: (user.phone ?? null) as string | null,
    contactUserId: null,
    contactFirstName: null,
    contactLastName: null,
    contactEmail: null,
    contactPhone: null,
  }
}

function payloadHash(snapshot: PartySnapshot): string {
  // Match Python's json.dumps(snapshot.__dict__, sort_keys=True, separators=(",", ":"))
  const sorted = Object.fromEntries(
    Object.entries(snapshot).sort(([a], [b]) => a.localeCompare(b))
  )
  const compact = JSON.stringify(sorted)
  return createHash('sha256').update(compact, 'utf-8').digest('hex')
}

function applySnapshot(
  event: BillingCustomerOutboxRow,
  snapshot: PartySnapshot,
  hash: string,
  now: number
): void {
  event.name = snapshot.name
  event.email = snapshot.email
  event.customerKind = snapshot.customerKind
  event.companyName = snapshot.companyName
  event.firstName = snapshot.firstName
  event.lastName = snapshot.lastName
  event.phone = snapshot.phone
  event.contactUserId = snapshot.contactUserId
  event.contactFirstName = snapshot.contactFirstName
  event.contactLastName = snapshot.contactLastName
  event.contactEmail = snapshot.contactEmail
  event.contactPhone = snapshot.contactPhone
  event.payloadHash = hash
  event.occurredAt = BigInt(now)
}

async function enqueueCustomerEnsure(
  deps: BillingCustomerSyncDeps,
  snapshot: PartySnapshot,
  now: number
): Promise<boolean> {
  const hash = payloadHash(snapshot)
  const latest = await deps.repository.findLatestOutboxBySubject(
    snapshot.subjectType,
    snapshot.subjectId
  )

  if (latest) {
    if (latest.status === 'pending' || latest.status === 'processing') {
      if (latest.payloadHash !== hash) {
        applySnapshot(latest, snapshot, hash, now)
        latest.updatedAt = BigInt(now)
        await deps.repository.updateOutboxEvent(latest.id, {
          name: latest.name,
          email: latest.email,
          customerKind: latest.customerKind,
          companyName: latest.companyName,
          firstName: latest.firstName,
          lastName: latest.lastName,
          phone: latest.phone,
          contactUserId: latest.contactUserId,
          contactFirstName: latest.contactFirstName,
          contactLastName: latest.contactLastName,
          contactEmail: latest.contactEmail,
          contactPhone: latest.contactPhone,
          payloadHash: latest.payloadHash,
          occurredAt: latest.occurredAt,
          updatedAt: latest.updatedAt,
        })
      }
      return false
    }

    if (latest.status === 'delivered' && latest.payloadHash === hash) {
      return false
    }
  }

  const event: BillingCustomerOutboxRow = {
    id: generateId('billingCustomerEvent'),
    eventType: CUSTOMER_EVENT_TYPE,
    subjectType: snapshot.subjectType,
    subjectId: snapshot.subjectId,
    name: snapshot.name,
    email: snapshot.email,
    occurredAt: BigInt(now),
    status: 'pending',
    attemptCount: 0,
    availableAt: BigInt(now),
    lockedAt: null,
    deliveredAt: null,
    lastError: null,
    createdAt: BigInt(now),
    updatedAt: BigInt(now),
    customerKind: null,
    companyName: null,
    firstName: null,
    lastName: null,
    phone: null,
    contactUserId: null,
    contactFirstName: null,
    contactLastName: null,
    contactEmail: null,
    contactPhone: null,
    payloadHash: null,
  }
  applySnapshot(event, snapshot, hash, now)
  await deps.repository.createOutboxEvent(event)
  return true
}

export async function enqueueCustomerEnsureForOrganization(
  deps: BillingCustomerSyncDeps,
  organization: OrganizationRow,
  now: number
): Promise<void> {
  const snapshot = await snapshotForOrganization(deps, organization)
  await enqueueCustomerEnsure(deps, snapshot, now)
}

export async function enqueueCustomerEnsureForUser(
  deps: BillingCustomerSyncDeps,
  user: UserRow,
  now: number
): Promise<void> {
  const snapshot = snapshotForUser(user)
  await enqueueCustomerEnsure(deps, snapshot, now)
}

export function customerEventPayload(
  event: BillingCustomerOutboxRow
): Record<string, unknown> {
  const common: Record<string, unknown> = {
    customerKind:
      event.customerKind ??
      (event.subjectType === 'organization' ? 'BUSINESS' : 'INDIVIDUAL'),
    name: event.name,
    email: event.email,
    firstName: event.firstName,
    lastName: event.lastName,
    phone: event.phone,
  }

  if (event.subjectType === 'organization') {
    return {
      ...common,
      customerType: 'CORE_ORGANIZATION',
      organizationId: event.subjectId,
      companyName: event.companyName,
      primaryContact: contactPayload(event),
    }
  }

  return {
    ...common,
    customerType: 'CORE_USER',
    userId: event.subjectId,
  }
}

function contactPayload(
  event: BillingCustomerOutboxRow
): Record<string, unknown> | null {
  if (!event.contactUserId) return null
  return {
    userId: event.contactUserId,
    firstName: event.contactFirstName,
    lastName: event.contactLastName,
    email: event.contactEmail,
    phone: event.contactPhone,
  }
}

export async function enqueueReconcileAll(
  deps: BillingCustomerSyncDeps,
  now: number
): Promise<{ organizations: number; users: number }> {
  const organizations = await deps.repository.listOrganizations()
  let organizationCount = 0
  for (const organization of organizations) {
    const snapshot = await snapshotForOrganization(deps, organization)
    if (await enqueueCustomerEnsure(deps, snapshot, now)) {
      organizationCount += 1
    }
  }

  const knownUserIds = await deps.repository.listKnownUserIds()
  let userCount = 0
  for (const userId of knownUserIds) {
    const user = await deps.repository.findUserById(userId)
    if (!user) continue
    const snapshot = snapshotForUser(user)
    if (await enqueueCustomerEnsure(deps, snapshot, now)) {
      userCount += 1
    }
  }

  return { organizations: organizationCount, users: userCount }
}

// Exported for tests — mirrors Python's private but tests import via snapshot helpers
export { payloadHash as _payloadHash }
